File: prolog/pilha.pl
% ---------------------------------------------------------
% Regras e Fatos Base (Lógica da Pilha)
% ---------------------------------------------------------
is_empty([]).
push(X, PilhaOriginal, [X | PilhaOriginal]).
pop([Topo | Resto], Topo, Resto).

% ---------------------------------------------------------
% Interface Interativa
% ---------------------------------------------------------
iniciar :-
    writeln('================================='),
    writeln('  SIMULADOR DE PILHA - PROLOG'),
    writeln('================================='),
    menu([]). % Inicia a recursão com a pilha (lista) vazia

menu(PilhaAtual) :-
    writeln('\nO que voce deseja fazer?'),
    writeln('1. Empilhar (Push)'),
    writeln('2. Desempilhar (Pop)'),
    writeln('3. Sair'),
    writeln('IMPORTANTE: No Prolog, toda entrada deve terminar com um PONTO (ex: 1. ou 10.)'),
    write('Opcao: '),
    read(Opcao),
    processar_opcao(Opcao, PilhaAtual).

% Se escolher 1, lê o valor, empilha e chama o menu de novo com a NovaPilha
processar_opcao(1, PilhaAtual) :-
    write('Digite o valor a empilhar (termine com "."): '),
    read(Valor),
    push(Valor, PilhaAtual, NovaPilha),
    format('\n[OK] ~w empilhado.~n', [Valor]),
    format('=> Estado atual da Pilha: ~w~n', [NovaPilha]),
    menu(NovaPilha).

% Se escolher 2 e a pilha estiver vazia
processar_opcao(2, []) :-
    writeln('\n[ERRO] A pilha esta vazia! Nao ha o que desempilhar.'),
    menu([]).

% Se escolher 2 e a pilha tiver itens, desempilha e chama o menu de novo com a NovaPilha
processar_opcao(2, PilhaAtual) :-
    pop(PilhaAtual, Desempilhado, NovaPilha),
    format('\n[OK] Valor desempilhado: ~w~n', [Desempilhado]),
    format('=> Estado atual da Pilha: ~w~n', [NovaPilha]),
    menu(NovaPilha).

% Se escolher 3, encerra a recursão e o programa fecha
processar_opcao(3, _) :-
    writeln('\nEncerrando a demonstracao...').

% Se digitar qualquer outra coisa
processar_opcao(_, PilhaAtual) :-
    writeln('\n[ERRO] Opcao invalida! Lembre-se do ponto no final.'),
    menu(PilhaAtual).

File: Java/pilha/src/Pilha.java
import java.util.Scanner;

public class Pilha {
    private int maxSize;
    private int[] stackArray;
    private int top;

    // Construtor
    public Pilha(int size) {
        this.maxSize = size;
        this.stackArray = new int[maxSize];
        this.top = -1;
    }

    // Comportamento: Empilhar
    public void push(int value) {
        if (top < maxSize - 1) {
            stackArray[++top] = value;
            System.out.println("\n[OK] " + value + " empilhado.");
        } else {
            System.out.println("\n[ERRO] Pilha cheia! Não é possível empilhar.");
        }
    }

    // Comportamento: Desempilhar
    public int pop() {
        if (!isEmpty()) {
            int value = stackArray[top--];
            System.out.println("\n[OK] Valor desempilhado: " + value);
            return value;
        } else {
            System.out.println("\n[ERRO] A pilha está vazia! Não há o que desempilhar.");
            return -1;
        }
    }

    // Verifica se está vazia
    public boolean isEmpty() {
        return (top == -1);
    }

    // Método auxiliar para visualizar a pilha no terminal
    public void exibir() {
        if (isEmpty()) {
            System.out.println("=> Estado atual da Pilha: []");
            return;
        }
        System.out.print("=> Estado atual da Pilha: [");
        for (int i = 0; i <= top; i++) {
            System.out.print(stackArray[i]);
            if (i < top) {
                System.out.print(", ");
            }
        }
        System.out.println("]");
    }  
}